src/pb_stub.cc

```diff
@@ -909,29 +932,196 @@ Stub::SendLogMessage(std::unique_ptr<PbLog>& log_send_message)
   std::unique_ptr<IPCMessage> log_request_msg =
       IPCMessage::Create(shm_pool_, false /* inline_response */);
   log_request_msg->Args() = log_request_shm->ShmHandle();
+  log_request_msg->Command() = PYTHONSTUB_LogRequest;
   ScopedDefer _([send_message_payload] {
     {
-      bi::scoped_lock<bi::interprocess_mutex> guard{
-          send_message_payload->log_mu};
+      bi::scoped_lock<bi::interprocess_mutex> guard{send_message_payload->mu};
       send_message_payload->waiting_on_stub = false;
-      send_message_payload->log_cv.notify_all();
+      send_message_payload->cv.notify_all();
     }
   });
 
   {
     // Send a message to be caught by the log monitor thread in python_be.cc
-    bi::scoped_lock<bi::interprocess_mutex> guard{send_message_payload->log_mu};
-    SendIPCLogMessage(log_request_msg);
+    bi::scoped_lock<bi::interprocess_mutex> guard{send_message_payload->mu};
+    SendIPCUtilsMessage(log_request_msg);
     while (!send_message_payload->waiting_on_stub) {
-      send_message_payload->log_cv.wait(guard);
+      send_message_payload->cv.wait(guard);
+    }
+  }
+}
+
+void
+Stub::SendCleanupId(void* id)
+{
+  std::unique_ptr<IPCMessage> ipc_message =
+      IPCMessage::Create(shm_pool_, true /* inline_response */);
+  ipc_message->Command() = PYTHONSTUB_CleanupRequest;
+  AllocatedSharedMemory<char> cleanup_request_message =
+      shm_pool_->Construct<char>(
+          sizeof(CleanupMessage) +
+          sizeof(bi::managed_external_buffer::handle_t));
+  CleanupMessage* cleanup_message_ptr =
+      reinterpret_cast<CleanupMessage*>(cleanup_request_message.data_.get());
+  cleanup_message_ptr->id = id;
+  cleanup_message_ptr->waiting_on_stub = false;
+  ipc_message->Args() = cleanup_request_message.handle_;
+
+  ScopedDefer _([&ipc_message, cleanup_message_ptr] {
+    {
+      bi::scoped_lock<bi::interprocess_mutex> lock{
+          *(ipc_message->ResponseMutex())};
+      cleanup_message_ptr->waiting_on_stub = false;
+      ipc_message->ResponseCondition()->notify_all();
+    }
+  });
```

Comment:
I was wondering why this additional cleanup step is required?

Reply:
Removed the additional round of communication.